### PR TITLE
Automatition

### PR DESCRIPTION
**Pull request by spaceh3ad**
- restructured repo for libs support
- added BETA price fetcher with UniswapV3 (need CR and further resolving)
- chainlink automation mocks
- unified solc version to 0.8.16

### CODE REVIEW

File: blockchain/src/YieldMate.sol [hunk 1]
```diff
@@ -1,8 +1,6 @@
 // SPDX-License-Identifier: UNLICENSED
-pragma solidity ^0.8.13;
+pragma solidity 0.8.16;
```
Comment (Mc01) [inline blockchain/src/YieldMate.sol:2]: Why not `0.8.20`?

File: blockchain/src/libs/order-manager/IOrderManager.sol [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    function addOrder(
+        address _token,
+        uint256 _amount,
+        uint256 _price
+    ) external returns (uint256);
+
+    function executeOrders() external;
+
+    function getEligbleOrders() external view;
+    // returns (OrderInfo[] memory);
```
Comment (Mc01) [inline blockchain/src/libs/order-manager/IOrderManager.sol:16]: Leftover to delete?

File: blockchain/src/libs/order-manager/OrderManager.sol [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    using Counters for Counters.Counter;
+    using EnumerableSet for EnumerableSet.UintSet;
+
+    PriceEngine private engine;
+
+    Counters.Counter orderId;
+
+    mapping(address => uint256) internal userToOrderMapping;
+    mapping(uint256 => OrderInfo) internal ordersMapping;
+    EnumerableSet.UintSet internal orders; // we store here orderIds'
```
[the remark this answers is not in the document]
Comment (Mc01) [reply blockchain/src/libs/order-manager/OrderManager.sol:22]: Potentially could work with struct returned on storage slot

File: blockchain/src/libs/order-manager/OrderManager.sol [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            assetOut: _assetOut,
+            amountIn: _amountIn,
+            status: OrderStatus({executed: false, amountOut: 0}),
+            orderType: _orderType
+        });
+
+        // add orderId to orders array
+        orders.add(_orderId);
+
+        orderId.increment();
```
Comment (Mc01) [inline blockchain/src/libs/order-manager/OrderManager.sol:64]: Just for clarity I would extract this into internal function and maybe even include in the Lib mentioned above

File: blockchain/src/libs/order-manager/lib/Objects.sol [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+    address assetOut;
+    uint256 amountIn;
+    OrderStatus status;
+    OrderType orderType;
+}
+
+enum OrderType {
+    BUY,
+    SELL
+}
```
[the remark this answers is not in the document]
Comment (Mc01) [reply blockchain/src/libs/order-manager/lib/Objects.sol:16]: Maybe something like orderId would be better to attach to particular order?

File: blockchain/src/libs/price-engine/chainlink-data-feeds/PriceConsumerV3.sol [hunk 6]
```diff
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.16;
+
+import "chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
```
Comment (Mc01) [inline blockchain/src/libs/price-engine/chainlink-data-feeds/PriceConsumerV3.sol:4]: Explicit import

File: blockchain/src/libs/price-engine/chainlink-data-feeds/PriceConsumerV3.sol [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+     * Returns the latest price.
+     */
+    function getLatestPrice() public view returns (int) {
+        // prettier-ignore
+        (
+            /* uint80 roundID */,
+            int price,
+            /*uint startedAt*/,
+            /*uint timeStamp*/,
+            /*uint80 answeredInRound*/
```
Comment (Mc01) [inline blockchain/src/libs/price-engine/chainlink-data-feeds/PriceConsumerV3.sol:30]: 💪

File: blockchain/src/libs/price-engine/PriceEngine.sol [hunk 8]
```diff
@@ -0,0 +1,58 @@
+//SPDX-License-Identifier: MIT
+pragma solidity 0.8.16;
+
+import "@uniswap/v3-core/contracts/interfaces/IUniswapV3Factory.sol";
+import "@uniswap/v3-core/contracts/interfaces/IUniswapV3Pool.sol";
+
+import "forge-std/console.sol";
```
Comment (Mc01) [inline blockchain/src/libs/price-engine/PriceEngine.sol:7]: Explicit imports (ps. don't include dev imports to GitHub 😉

File: blockchain/src/libs/price-engine/PriceEngine.sol [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        0x1F98431c8aD98523631AE4a59f267346ea31F984;
+    uint24 public fee = 3000;
+
+    function getPrice()
+        external
+        view
+        returns (
+            // address tokenIn,
+            // address tokenOut
+            uint256
```
Comment (Mc01) [inline blockchain/src/libs/price-engine/PriceEngine.sol:24]: Not sure 😅 Natspec?

File: blockchain/src/libs/price-engine/PriceEngine.sol [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        price = price * 10 ** 18; // TODO: czemu do 18? 18-6 => 12
+        console.log(price);
+        if (price < Q96) {
+            price = Q96 / price;
+        } else {
+            price = price / Q96;
+        }
+
+        console.log(price);
+        // TODO: jak wykminic co tu dostajemy? price dla WMATIC/USDC czy USDC/WMATIC
```
Comment (Mc01) [inline blockchain/src/libs/price-engine/PriceEngine.sol:54]: Not better to just use OracleLibrary and TWAP for that?

File: blockchain/src/mocks/MockToken.sol [hunk 11]
```diff
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: UNLICENSED
-pragma solidity ^0.8.9;
+pragma solidity 0.8.16;
 
-import "@openzeppelin/token/ERC20/ERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
```
Comment (Mc01) [inline blockchain/src/mocks/MockToken.sol:4]: Explicit imports

File: blockchain/test/libs/OrderManager.t.sol [hunk 12]
```diff
@@ -0,0 +1,52 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity 0.8.16;
+
+import "forge-std/Test.sol";
+import "forge-std/StdUtils.sol";
+
+import "../../src/libs/order-manager/OrderManager.sol";
+import "../../src/libs/price-engine/PriceEngine.sol";
```
Comment (Mc01) [inline blockchain/test/libs/OrderManager.t.sol:8]: Explicit imports

File: blockchain/test/libs/OrderManager.t.sol [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+        uint256 orderId = manager.addOrder(
+            USDC,
+            9000000,
+            10 * 10 ** 6,
+            WMATIC,
+            OrderType.BUY
+        );
+
+        assertEq(orderId, 0);
+    }
```
Comment (Mc01) [inline blockchain/test/libs/OrderManager.t.sol:51]: Could you add more comments to highlight the test cases more?

**Review by Mc01 — COMMENTED** [review]
Good wip 🖖

**Review by Mc01 — COMMENTED** [review]
Good wip 🖖

File: blockchain/src/YieldMate.sol (hunk 1 above)
Comment (spaceh3ad) [reply blockchain/src/YieldMate.sol:2]: doesnt work for me :shrug:

File: blockchain/src/libs/price-engine/PriceEngine.sol (hunk 10 above)
Comment (spaceh3ad) [reply blockchain/src/libs/price-engine/PriceEngine.sol:54]: that requires 0.7.6 solc, which conflicts with other versions we have

File: blockchain/src/YieldMate.sol (hunk 1 above)
Comment (spaceh3ad) [reply blockchain/src/YieldMate.sol:2]: foundry nie wspiera 0.8.20 XD
https://github.com/foundry-rs/foundry/issues/4933

**Pull request merged**